cdcv2/metadata/model.go

```diff
@@ -14,129 +14,217 @@
+}
+
+// Scan implements the sql.Scanner interface.
+func (cp *CaptureProgress) Scan(value interface{}) error {
+	b, ok := value.([]byte)
+	if !ok {
+		return errors.New("type assertion to []byte failed")
+	}
+
+	return json.Unmarshal(b, cp)
+}
 
 // SchedState is the type of state to schedule owners and processors.
 type SchedState int
 
 const (
+	// SchedInvalid should never be used.
+	SchedInvalid SchedState = SchedState(0)
 	// SchedRemoved means the owner or processor is removed.
-	SchedRemoved SchedState = SchedState(0)
+	SchedRemoved SchedState = SchedState(1)
 	// SchedLaunched means the owner or processor is launched.
-	SchedLaunched SchedState = SchedState(1)
+	SchedLaunched SchedState = SchedState(2)
 	// SchedRemoving means the owner or processor is in removing.
-	SchedRemoving SchedState = SchedState(2)
-
-	totalStates int = 3
+	SchedRemoving SchedState = SchedState(3)
 )
 
-// ScheduledChangefeed is for owner and processor schedule.
-type ScheduledChangefeed struct {
-	ChangefeedID ChangefeedIDWithEpoch
-	CaptureID    model.CaptureID
-	State        SchedState
+// String implements the fmt.Stringer interface.
+func (s SchedState) String() string {
+	str, err := s.toString()
+	if err != nil {
+		log.Error("failed to convert SchedState to string", zap.Error(err))
+	}
+	return str
+}
+
+func (s SchedState) toString() (string, error) {
+	switch s {
+	case SchedLaunched:
+		return "Launched", nil
+	case SchedRemoving:
+		return "Removing", nil
+	case SchedRemoved:
+		return "Removed", nil
+	default:
+		return "unreachable", nil
+	}
+}
+
+func (s SchedState) fromString(str string) error {
+	switch str {
+	case "Launched":
+		s = SchedLaunched
+	case "Removing":
+		s = SchedRemoving
+	case "Removed":
+		s = SchedRemoved
+	default:
+		s = SchedInvalid
+		return errors.New("unreachable")
+	}
+	return nil
+}
+
+// Value implements the driver.Valuer interface.
+func (s SchedState) Value() (driver.Value, error) {
+	return s.toString()
+}
 
-	// TaskPosition is used for creating owner and processors on captures.
-	//
-	// When controller specifies resources to a changefeed, it won't care about it.
-	TaskPosition ChangefeedProgress
+// Scan implements the sql.Scanner interface.
+func (s *SchedState) Scan(value interface{}) error {
+	b, ok := value.([]byte)
+	if !ok {
+		return errors.New("type assertion to []byte failed")
+	}
+
+	return s.fromString(string(b))
 }
 
-// ChangefeedSchedule is used to query changefeed schedule information.
-type ChangefeedSchedule struct {
-	Owner      ScheduledChangefeed
-	Processors []ScheduledChangefeed
+// ScheduledChangefeed is for owner and processor schedule.
+type ScheduledChangefeed struct {
+	ChangefeedUUID ChangefeedUUID   `gorm:"column:changefeed_uuid;type:bigint(20) unsigned;primaryKey" json:"changefeed_uuid"`
+	Owner          *model.CaptureID `gorm:"column:owner;type:varchar(128)" json:"owner"`
+	OwnerState     SchedState       `gorm:"column:owner_state;type:text;not null" json:"owner_state"`
```

Comment:
use `enum` or varchar?

Reply:
We will handle these changes in another PR.